app/src/main/kotlin/com/wire/android/ui/home/conversations/MessageExpiration.kt

```diff
@@ -0,0 +1,266 @@
+import androidx.compose.runtime.remember
+import androidx.compose.runtime.setValue
+import androidx.compose.ui.platform.LocalContext
+import com.wire.android.R
+import com.wire.android.ui.home.conversations.model.ExpirationStatus
+import com.wire.android.ui.home.conversations.model.UIMessage
+import com.wire.android.ui.home.conversations.model.UIMessageContent
+import com.wire.kalium.logic.data.message.Message
+import kotlinx.coroutines.delay
+import kotlinx.datetime.Clock
+import kotlin.time.Duration
+import kotlin.time.Duration.Companion.ZERO
+import kotlin.time.Duration.Companion.days
+import kotlin.time.Duration.Companion.hours
+import kotlin.time.Duration.Companion.minutes
+import kotlin.time.Duration.Companion.seconds
+
+@Composable
+fun rememberSelfDeletionTimer(expirationStatus: ExpirationStatus): SelfDeletionTimer.SelfDeletionTimerState {
+    val context = LocalContext.current
+
+    return remember { SelfDeletionTimer(context).fromExpirationStatus(expirationStatus) }
+}
+
+class SelfDeletionTimer(private val context: Context) {
+
+    fun fromExpirationStatus(expirationStatus: ExpirationStatus): SelfDeletionTimerState {
+        return if (expirationStatus is ExpirationStatus.Expirable) {
+            with(expirationStatus) {
+                val timeLeft = calculateTimeLeft(selfDeletionStatus, expireAfter)
+                SelfDeletionTimerState.Expirable(context.resources, timeLeft, expireAfter)
+            }
+        } else {
+            SelfDeletionTimerState.NotExpirable
+        }
+    }
+
+    private fun calculateTimeLeft(
+        selfDeletionStatus: Message.ExpirationData.SelfDeletionStatus?,
+        expireAfter: Duration
+    ): Duration {
+        return if (selfDeletionStatus is Message.ExpirationData.SelfDeletionStatus.Started) {
+            val timeElapsedSinceSelfDeletionStartDate = Clock.System.now() - selfDeletionStatus.selfDeletionStartDate
+            val timeLeft = expireAfter - timeElapsedSinceSelfDeletionStartDate
+
+            // time left for deletion it can be a negative value if the time difference between the self deletion start date and
+            // now is greater then expire after millis, we normalize it to 0 seconds
+            if (timeLeft.isNegative()) {
+                ZERO
+            } else {
+                timeLeft
+            }
+        } else {
+            expireAfter
+        }
+    }
+
+    sealed class SelfDeletionTimerState {
+
+        class Expirable(
+            private val resources: Resources,
+            timeLeft: Duration,
+            private val expireAfter: Duration
+        ) : SelfDeletionTimerState() {
+            companion object {
+                private const val TIME_LEFT_RATIO_BOUNDARY_FOR_ALMOST_TIME_LEFT_ELAPSED_ALPHA = 0.75
+                private const val PAST_RATIO_BOUNDARY_FOR_ALMOST_TIME_LEFT_ALPHA_VALUE = 0F
+                private const val BEFORE_RATIO_BOUNDARY_FOR_ALMOST_TIME_LEFT_ALPHA_VALUE = 1F
```

Comment:
What are these supposed to mean exactly?

What is "almost time left"?

Care to please add some docs?

Reply:
I guess I'll never know 😅